### PR TITLE
Regression tests: Algorithm

### LINKED ISSUE #2913: Add comparison tests for Algorithm validation in new validation model
Need to add tests to compare the new validation path against the existing one to ensure no regressions are introduced in the new path and that all scenarios behave as expected.

- Add tests comparing `ValidateTokenAsync` in the new path against the existing one for Algorithm validation scenarios.

Part of #2711

### PR DESCRIPTION
**Pull request by iNinja**
# Regression tests: Algorithm
- Added AlgorithmValidationError
- Refactored error creation in ValidateSignature to surface the invalid algorithm as part of the error
- Added regression/comparison tests for algorithm validation scenarios

Resolves #2913 

Part of #2711

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs [hunk 1]
```diff
@@ -241,13 +241,29 @@ private static ValidationResult<SecurityKey> ValidateSignatureWithKey(
...
+                        new MessageDetail(
+                            TokenLogMessages.IDX10518,
+                            algorithmValidationError.MessageDetail.Message),
+                        typeof(SecurityTokenInvalidAlgorithmException),
+                        new StackFrame(true),
+                        algorithmValidationError.InvalidAlgorithm);
+                }
+                else
+                {
+                    // overridden delegate did not return an AlgorithmValidationError
```
Comment (kllysng) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:257]: Tests: do you plan on adding test coverage for this else block?

Comment (iNinja) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:257]: We would have look at this in our extensibility scenarios. We are adding specific tests for the cases where the delegates are implemented by users.

Comment (kllysng) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:257]: Thank you, resolving this!

**Review by kllysng — APPROVED** [review]
Thanks, @iNinja

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Algorithm.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                theoryData.Add(new ValidateTokenAsyncAlgorithmTheoryData("Valid_ValidAlgorithmsIsEmptyList")
+                {
+                    SigningCredentials = KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2,
+                    TokenValidationParameters = CreateTokenValidationParameters(
+                        KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2.Key, validAlgorithms: []),
+                    ValidationParameters = CreateValidationParameters(
+                        KeyingMaterial.DefaultX509SigningCreds_2048_RsaSha2_Sha2.Key, validAlgorithms: []),
+                });
+
+                theoryData.Add(new ValidateTokenAsyncAlgorithmTheoryData("Invalid_TokenIsSignedWithAnInvalidAlgorithm")
```
Comment (westin-m) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Algorithm.cs:64]: **Question**: Is this the only negative test case that's needed?

**Review by jennyf19 — APPROVED** [review]

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Algorithm.cs (hunk 2 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Algorithm.cs:64]: Yes. The algorithm validation fails **only** when **all** the following conditions are met:
- ValidAlgorithms is not null
- ValidAlgorithms is not empty
- The algorithm used is **not** present in the ValidAlgorithms list

Any other scenario is considered valid. I intentionally added tests highlighting those to help explain this.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs [hunk 3]
```diff
@@ -241,13 +241,29 @@ private static ValidationResult<SecurityKey> ValidateSignatureWithKey(
...
             if (!result.IsValid)
-                return new ValidationError(
-                    new MessageDetail(
-                        TokenLogMessages.IDX10518,
-                        result.UnwrapError().MessageDetail.Message),
-                    ValidationFailureType.SignatureAlgorithmValidationFailed,
-                    typeof(SecurityTokenInvalidAlgorithmException),
-                    new StackFrame(true));
+            {
+                if (result.UnwrapError() is AlgorithmValidationError algorithmValidationError)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs:245]: Curious as to why we need to recreate the AlgorithmValidationError?

**Pull request merged**